src/ProxyKit.Tests/RoundRobinTests.cs

```diff
@@ -48,5 +49,22 @@ public void Given_two_hosts_with_weight_then_round_robin_should_distribute_evenl
             upstreamHost = roundRobin.Next();
             upstreamHost.ToString().ShouldBe("http://localhost:5001/");
         }
+
+        [Fact]
+        public void Given_one_host_then_ensure_lock_is_released()
+        {
+            var roundRobin = new RoundRobin
+            {
+                new UpstreamHost("http://localhost:5001/")
+            };
+            var upstreamHost = roundRobin.Next();
+            upstreamHost.ToString().ShouldBe("http://localhost:5001/");
+            var thread = new Thread(() =>
+            {
+                roundRobin.Add(new UpstreamHost("http://localhost:5002/"));
```

Comment:
This would attempt to take write lock when adding from another thread. The write lock would never be successfully be taken because the read was never given back.